nixos/modules/services/misc/klipper.nix

```diff
@@ -0,0 +1,59 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.klipper;
+  package = pkgs.klipper;
+  format = pkgs.formats.ini { mkKeyValue = generators.mkKeyValueDefault {} ":"; };
+in
+{
+  ##### interface
+  options = {
+    services.klipper = {
+      enable = mkEnableOption "Klipper, the 3D printer firmware";
+
+      octoprintIntegration = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Allows Octoprint to control Klipper.";
+      };
+
+      settings = mkOption {
+        type = format.type;
+        default = { };
+        description = ''
+          Configuration for Klipper. See the <link xlink:href="https://www.klipper3d.org/Overview.html#configuration-and-tuning-guides">documentation</link>
+          for supported values.
+        '';
+      };
+    };
+  };
+
+  ##### implementation
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.octoprintIntegration -> config.services.octoprint.enable;
+      message = "Option services.klipper.octoprintIntegration requires services.octoprint.";
```

Comment:
Maybe write that octoprint must be enabled  and then the options name.

Also does this support octoprint from another machine?

Reply:
It does not, they must be running on the same machine. The way klipper works is by exposing a faux pts in `/run/klipper/tty` which octoprint controls as if it were the printer controller. Klipper then performs most of the complex computations (microstepping, etc) and sends dump position data to the actual printer controller.

I rephrased the assertion.